File: used_py/LCMOEA/problem.py
import numpy as np

class Problem:

    def __init__(self, objectives, dim, UBLB, equality_constraints, inequality_constraints, www):
        self.objectives = objectives  # 目标函数列表
        self.num_obj = len(objectives)  # 目标函数的数量
        self.dim = dim  # 决策变量的数量
        if UBLB.shape[0] == 2 and UBLB.size == 2:
            self.lb = UBLB[1] * np.ones(dim)
            self.ub = UBLB[0] * np.ones(dim)
        elif UBLB.shape[0] == 2 and UBLB.shape[1] == dim:
            self.lb = UBLB[1, :]
            self.ub = UBLB[0, :]
        else:
            raise ValueError("UBLB 的形状不正确。应为 (2,dim) 或 (2, 1)。")
        self.equality_constraints = equality_constraints  # 等式约束条件列表
        self.inequality_constraints = inequality_constraints  # 不等式约束条件列表
        self.num_constr = len(equality_constraints) + len(inequality_constraints)  # 约束条件的数量
        # if UBLB.shape[0] == 2 and UBLB.size == 2:
        #     self.lb = UBLB[0] * np.ones(population.shape[1])
        #     self.ub = UBLB[1] * np.ones(population.shape[1])
        # elif UBLB.shape[0] == 2 and UBLB.shape[1] == population.shape[1]:
        #     self.lb = UBLB[0, :]
        #     self.ub = UBLB[1, :]
        # else:
        #     raise ValueError("UBLB 的形状不正确。应为 (2,dim) 或 (2, 1)。")
        self.www = www#平衡数量级
    
    def evaluate(self, x):
        obj_w = []
        for obj_fun in self.objectives:
            obj_w.append(obj_fun(x))
        obj_w = np.array(obj_w)
        obj_w = np.dot(obj_w, self.www)
        return obj_w

File: old_matlab/matlab1/Problem.m
classdef Problem
    properties
        objectives  % 目标函数列表
        num_obj  % 目标函数数量
        dim  % 决策变量数量
        lb  % 下界
        ub  % 上界
        equality_constraints  % 等式约束列表
        inequality_constraints  % 不等式约束列表
        num_constr  % 约束数量
        www  % 目标函数的权重向量
    end

    methods
        function obj = Problem(objectives, dim, UBLB, equality_constraints, inequality_constraints, www)
            obj.objectives = objectives;
            obj.num_obj = length(objectives);
            obj.dim = dim;
            if size(UBLB, 1) == 2 && numel(UBLB) == 2
                obj.lb = UBLB(2) * ones(dim, 1);
                obj.ub = UBLB(1) * ones(dim, 1);
            elseif size(UBLB, 1) == 2 && size(UBLB, 2) == dim
                obj.lb = UBLB(2, :)';
                obj.ub = UBLB(1, :)';
            else
                error('UBLB 的形状不正确。应为 (2,dim) 或 (2, 1)。');
            end
            obj.equality_constraints = equality_constraints;
            obj.inequality_constraints = inequality_constraints;
            obj.num_constr = length(equality_constraints) + length(inequality_constraints);
            obj.www = www;
        end

        function obj_w = evaluate(obj, x)
            obj_values = zeros(obj.num_obj, 1);
            for i = 1:obj.num_obj
                obj_values(i) = obj.objectives{i}(x);
            end
            obj_w = obj_values' * obj.www;
        end
    end
end
